doc/_extensions/zephyr/link-roles.py

```diff
@@ -68,16 +47,49 @@ def setup(app):
     }
 
 
-def autolink(pattern):
+def modulelink(default_module=None, format="blob"):
     def role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+        # Set default values
+        module = default_module
+        rev = get_github_rev()
+        baseurl = inliner.document.settings.env.app.config.link_roles_default_url
+
         m = re.search(r"(.*)\s*<(.*)>", text)
         if m:
             link_text = m.group(1)
             link = m.group(2)
         else:
             link_text = text
             link = text
-        url = pattern % (link,)
+
+        module_match = re.search(r"(.+?):\s*(.+)", link)
+        if module_match:
+            module = module_match.group(1)
+            link = module_match.group(2)
+
+        # Try to get a module repository's GitHub URL from the manifest.
+        #
+        # This allows e.g. building the docs in downstream Zephyr-based
+        # software with forks of the zephyr repository, and getting
+        # :zephyr_file: / :zephyr_raw: output that links to the fork,
+        # instead of mainline zephyr.
+        projects = [p.name for p in west_manifest.projects] if west_manifest else []
+        if module in projects:
+            project = west_manifest.get_projects([module])[0]
+            baseurl = project.url
+            rev = project.revision
+
+        if baseurl is None or module is None:
```

Comment:
can this condition ever hold true?

Reply:
`module` can be `None` if using the new role `:module_file:` without a module argument. 
`baseurl` can be `None` given the following circumstances:
- `:module_file:` is used while the `link_roles_default_url` config value is not set (defaults to `None`). 
- The provided module is not found in the manifest (invalid module)

Reply:
Failure tests rewritten to include the case where both a default url and an invalid module are provided